### PR TITLE
Upgrade to TypeChain v2

### PR DESCRIPTION
**Pull request by nlordell**
This PR updates the TypeChain targets to the v2 versions. This will hopefully keep us getting TypeChain improvements. For example, the `Linkable` type and cast is no longer needed :).

### Test Plan

CI, this should only affect generated typings and contains no real code modifications (besides adaptations to use new types).

### CODE REVIEW

**Review by e00E — APPROVED** [review]

**Review by fleupold — APPROVED** [review]
Nice, less custom types are great!

File: test/BatchExchange.spec.ts [hunk 1]
```diff
@@ -72,14 +71,8 @@ contract("BatchExchange", async (accounts) => {
...
-    await (BatchExchange as Linkable<BatchExchangeInstance>).link(
-      "IdToAddressBiMap",
-      lib1.address,
-    );
-    await (BatchExchange as Linkable<BatchExchangeInstance>).link(
-      "IterableAppendOnlySet",
-      lib2.address,
-    );
+    await BatchExchange.link("IdToAddressBiMap", lib1.address);
+    await BatchExchange.link("IterableAppendOnlySet", lib2.address);
```
Comment (bh2smith) [inline test/BatchExchange.spec.ts:75]: Cool! I am happy about this!

File: test/BatchExchange.spec.ts [hunk 2]
```diff
@@ -166,9 +159,7 @@ contract("BatchExchange", async (accounts) => {
       );
     });
     it("Burns 10 OWL when adding token", async () => {
-      const TokenOWLProxy = artifacts.require(
-        "../node_modules/@gnosis.pm/owl-token/build/contracts/TokenOWLProxy",
-      );
+      const TokenOWLProxy = artifacts.require("TokenOWLProxy");
```
Comment (bh2smith) [inline test/BatchExchange.spec.ts:162]: Weird, I hadn't even noticed this was a thing. What was making the OWL token proxy so special before?

File: types/truffle-typings/index.d.ts [hunk 3]
```diff
@@ -2,10 +2,14 @@
...
 
-export interface Linkable<T> extends Truffle.Contract<T> {
-  link(lib: string, address: string): Promise<void>;
+declare global {
+  namespace Truffle {
+    interface Artifacts {
+      require(name: string): Truffle.Contract<unknown>;
+    }
+  }
 }
```
Comment (bh2smith) [inline types/truffle-typings/index.d.ts]: I guess this is going to conflict with my deposit migration #802. Should be easy enough to deal with the conflict on either side.

**Review by bh2smith — APPROVED** [review]
Awesome stuff!

File: test/BatchExchange.spec.ts (hunk 2 above)
Comment (nlordell) [reply test/BatchExchange.spec.ts:162]: Nothing, Truffle works with both methods. The main difference is that when you do it this way it works better with the generated types.

File: types/truffle-typings/index.d.ts (hunk 3 above)
Comment (nlordell) [reply types/truffle-typings/index.d.ts]: Trivial to merge.

**Pull request merged**